### ISSUE TITLE
Low mAP when training YOLOv8s on VisDrone?

### ISSUE THREAD

**Issue by davidkahle** [ultralytics/ultralytics#4047]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I trained a YOLOv8s model on VisDrone by running the command shown in the Ultralytics VisDrone documentation, only I set the number of epochs to 300 and the imgsz to 800. When I asked on other forums for advice about improving accuracy, many people told me the values I got are far too low -- mAP50 of 0.47 and mAP50-95 of 0.29

The only thing I have to compare to is the results found by [researchers in this paper](https://www.mdpi.com/2079-9292/12/10/2323) (see Table 2), but I was wondering if there's something I can do to improve accuracy? I know I could train with a larger model or greater image size, but this model will be running on a Nvidia Jetson Nano, so I don't want it to be too slow. Here's a screenshot of the output when I finished: 
![Selection_058](https://github.com/ultralytics/ultralytics/assets/66802522/9db9b242-e666-4836-bddc-6bc93fe29e86)

**Comment by glenn-jocher** [comment]
@davidkahle hello, thank you for raising this question.

It seems you've already trained your YOLOv8s model with an adequate number of epochs (300) and a full-size image (800). Despite these good practices, achieving high levels of accuracy can sometimes be complex due to the nature of the dataset and its variance.

Given the constraints you're experiencing with your Nvidia Jetson Nano, one approach could involve optimizing hyperparameters. This can help improve performance without increasing the computational requirements of the model drastically.

Exploring data augmentation techniques could also help to improve model performance. The model might perform better as it sees different variations of your data.

Another possible direction to investigate is if your dataset is evenly distributed among the classes. Imbalanced datasets can lead to lower accuracy rates, especially for minority classes.

Lastly, remember that typically larger YOLO models should provide better accuracy at the cost of increased complexity and computational requirements. Depending on your application, a trade-off between accuracy and speed might be necessary.

I hope this gives you some ideas for improving the accuracy of your model. Please let us know how you get on, or if you need any further advice.

**Comment by Zhier123** [comment]
Hi @glenn-jocher ,thank you for your , I have the same prob: yolov8s trained on dataset visdroneDET-2019, I also tried yolov8s-p2 and modified the yolov8s-p2.yaml and some other file adding Context Aggreaction Block after the head of yolov8s-p2 to imporve the performance of yolov8-p2 on this dataset. 
  These is What I have done to improve yolov8s' performance on visdroneDET-2019
  1.  use yolov8s-p2 instead of yolov8s
      > model = YOLO('yolov8s-p2_CA.yaml').load('yolov8s.pt')
      > model.train(data='VisDrone.yaml', epochs=100, imgsz=640)
  2. modified `nn.module.block.py`, `nn.module.__init__.py`, `nn.tasks.py` to add Context Aggretation Block and regist is.
  3. add file `yolov8s-p2_CA.yaml`  under  dir `cfg/models/v8/`, I adde ContextAggretation in these areas, and the rest of the file is same as yolov8s-p2.yaml.
  ~~~
  # YOLOv8.0-p2 head
head:
  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 6], 1, Concat, [1]]  
  - [-1, 3, C2f, [512]]  # 12

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 4], 1, Concat, [1]]  
  - [-1, 3, C2f, [256]]  # 15 (P3/8-small)

  - [-1, 1, nn.Upsample, [None, 2, 'nearest']]
  - [[-1, 2], 1, Concat, [1]]  
  - [-1, 3, C2f, [128]]  
  - [-1, 1, ContextAggregation, [128]] #19

  - [-1, 1, Conv, [128, 3, 2]]
  - [[-1, 15], 1, Concat, [1]]  
  - [-1, 3, C2f, [256]]  
  - [-1, 1, ContextAggregation, [256]] #23

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 12], 1, Concat, [1]]  
  - [-1, 3, C2f, [512]] 
  - [-1, 1, ContextAggregation, [512]] # 27

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 9], 1, Concat, [1]]  # cat head P5
  - [-1, 3, C2f, [1024]]  
  - [-1, 1, ContextAggregation, [1024]]  #31
                  
  - [[19, 23, 27, 31], 1, Detect, [nc]]  # Detect(P2, P3, P4, P5)

  ~~~
After doing these modifications, I achieved mAP50: 0.428 and mAP50-95: 0.258 after training for 60 epochs. However, the score is still too low. Like @davidkahle, I also have the constraint of using a relatively small model, so using yolovm or larger models may not be the best choice for me.

Could you please give me some advice on optimizing hyperparameters like imgsize or nms iou?

**Comment by Zhier123** [comment]
Hi @glenn-jocher ,thank you for your help , I have the same problem: yolov8s trained on dataset visdroneDET-2019, I also tried yolov8s-p2 and modified the yolov8s-p2.yaml and some other files adding Context Aggreaction Block after the `head` of yolov8s-p2 to imporve the performance of yolov8-p2 on this dataset.
These is What I have done to improve yolov8s' performance on visdroneDET-2019:

use yolov8s-p2 instead of yolov8s
- model = YOLO('yolov8s-p2_CA.yaml').load('yolov8s.pt')
- model.train(data='VisDrone.yaml', epochs=100, imgsz=640)

modified `nn.module.block.py`, `nn.module.__init__.py`, `nn.tasks.py` to add Context Aggretation Block and regist is.
add file `yolov8s-p2_CA.yaml` under dir `cfg/models/v8/`, I added ContextAggretation in the areas below, and the rest of the file is same as yolov8s-p2.yaml.
# YOLOv8.0-p2 head
head:
- [-1, 1, nn.Upsample, [None, 2, 'nearest']]
- [[-1, 6], 1, Concat, [1]]  
- [-1, 3, C2f, [512]]  # 12

- [-1, 1, nn.Upsample, [None, 2, 'nearest']]
- [[-1, 4], 1, Concat, [1]]  
- [-1, 3, C2f, [256]]  # 15 (P3/8-small)

- [-1, 1, nn.Upsample, [None, 2, 'nearest']]
- [[-1, 2], 1, Concat, [1]]  
- [-1, 3, C2f, [128]]  
- [-1, 1, ContextAggregation, [128]] #19

- [-1, 1, Conv, [128, 3, 2]]
- [[-1, 15], 1, Concat, [1]]  
- [-1, 3, C2f, [256]]  
- [-1, 1, ContextAggregation, [256]] #23

- [-1, 1, Conv, [256, 3, 2]]
- [[-1, 12], 1, Concat, [1]]  
- [-1, 3, C2f, [512]] 
- [-1, 1, ContextAggregation, [512]] # 27

- [-1, 1, Conv, [512, 3, 2]]
- [[-1, 9], 1, Concat, [1]]  # cat head P5
- [-1, 3, C2f, [1024]]  
- [-1, 1, ContextAggregation, [1024]]  #31
                
- [[19, 23, 27, 31], 1, Detect, [nc]]  # Detect(P2, P3, P4, P5)

After doing these modifications, I achieved mAP50: 0.428 and mAP50-95: 0.258 after training for 60 epochs. However, the score is still too low. Like @davidkahle, I also have the constraint of using a relatively small model, so using yolovm or larger models may not be the best choice for me.

Could you please give me some advice on optimizing hyperparameters like imgsize or nms iou?

**Comment by glenn-jocher** [comment]
@Zhier123 hello, it's great to hear that you're leveraging YOLOv8 and exploring ways to optimize it for your specific needs.

From your modifications, it seems that you've already made significant efforts to improve YOLOv8s's performance on the VisDroneDET-2019 dataset. Your implemented Context Aggregation Block might already help your model perform better on the given task.

Now, with regards to optimizing hyperparameters, it's important to remember that they can greatly impact the overall performance of the model. Below, you can find several recommendations:

1. Image Size (imgsz): If you increase the image size, the model usually sees more details, which improves the detection of small objects. However, this can also slow down the training process and require more computational resources. Try adjusting this up and down to find the right balance for your application.

2. Non-Maximum Suppression IoU (NMS IoU): Adjusting the Intersection over Union (IoU) threshold in non-maximum suppression can help to reduce false positives. A higher IoU value is more strict, meaning fewer bounding boxes will be selected.

3. Batch Size: If you're experiencing out-of-memory issues, consider reducing the batch size. Conversely, if your hardware resources allow, increasing the batch size could improve the stability of your training and potentially lead to better results.

4. Learning Rate: The learning rate is one of the most impactful hyperparameters. If your model is learning too slowly or too quickly, adjusting the learning rate up or down might be beneficial.

5. Augmentation Parameters: Data augmentation can help the model generalize from the training set to the test set. You might want to experiment with different types of augmentations and their levels.

Lastly, remember that these choices over hyperparameters can influence each other and the model outcome, so it might take some experimentation to find the best set for your specific need. I encourage you to try different combinations and evaluate how they impact your model's performance.
